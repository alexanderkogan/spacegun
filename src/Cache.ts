export class Cache<S, T> {
    private data: Map<S, { value: T, age: number }> = new Map()

    public constructor(
        public maxAgeSeconds: number,
        public maxSize: number = 100
    ) { }

    public get(key: S): T | undefined {
        const entry = this.data.get(key)
        if (entry !== undefined && Date.now() - entry.age < this.maxAgeSeconds * 1000) {
            return entry.value
        }
        return undefined
    }

    public set(key: S, value: T): void {
        this.data.set(key, { value, age: Date.now() })
    }

    public async calculate(key: S, valueProvider: () => Promise<T>): Promise<T> {
        let value = this.get(key)
        if (value === undefined) {
            value = await valueProvider()
            this.set(key, value)
        }
        return value
    }
}
