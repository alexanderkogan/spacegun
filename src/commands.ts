import { ClusterRepository } from "@/cluster/ClusterRepository"
import { Deployment } from "@/cluster/model/Deployment"

import { ImageRepository } from "@/images/ImageRepository"

import { pad } from "@/pad"
import chalk from "chalk"

import * as ora from "ora"
import { Question } from "@/Question"

export type Command = "pods" | "images" | "deployments" | "deploy" | "scalers" | "help"

async function load<T>(p: Promise<T>): Promise<T> {
    const progress = ora()
    progress.start("loading")
    const result = await p
    progress.stop()
    return result
}

async function foreachCluster(clusterRepository: ClusterRepository, f: (clusterRepository: ClusterRepository, cluster: string) => void) {
    for (const cluster of clusterRepository.clusters) {
        console.log("")
        console.log(chalk.underline.bold(pad(`${cluster}`)))
        await f(clusterRepository, cluster)
    }
}

async function podsCommand(clusterRepository: ClusterRepository, cluster: string) {
    const pods = await load(clusterRepository.pods(cluster))
    console.log(chalk.bold(pad("pod name", 5) + pad("tag name", 5) + pad("starts", 1), pad("status", 1)))
    pods.forEach(pod => {
        let restartText
        if (pod.restarts === undefined) {
            restartText = chalk.bold.cyan(pad("na!", 1))
        } else if (pod.restarts > 30) {
            restartText = chalk.bold.cyan(pad(pod.restarts.toString() + "!", 1))
        } else if (pod.restarts > 10) {
            restartText = chalk.bold.magenta(pad(pod.restarts.toString(), 1))
        } else {
            restartText = pad(pod.restarts.toString(), 1)
        }
        const statusText = pod.ready ? chalk.bold.magenta(pad("up", 1)) : chalk.bold.cyan(pad("down!", 1))
        let tagText
        if (pod.image === undefined) {
            tagText = chalk.bold.magenta(pad("missing", 5))
        } else {
            tagText = pad(pod.image.tag, 5)
        }
        console.log(pad(pod.name, 5) + tagText + restartText + statusText)
    })
}

async function imagesCommand(imageRepository: ImageRepository) {
    const images = await imageRepository.images()
    images.forEach(image =>
        console.log(image)
    )
}

function logDeploymentHeader() {
    console.log(chalk.bold(pad("deployment name", 5) + pad("tag name", 7)))
}

function logDeployment(deployment: Deployment) {
    let tagText
    if (deployment.image === undefined) {
        tagText = chalk.bold.magenta(pad("missing", 7))
    } else {
        tagText = pad(deployment.image.tag, 7)
    }
    console.log(pad(deployment.name, 5) + tagText)
}

async function deployementsCommand(clusterRepository: ClusterRepository, cluster: string) {
    const deployments = await load(clusterRepository.deployments(cluster))
    logDeploymentHeader()
    deployments.forEach(deployment => {
        logDeployment(deployment)
    })
}

async function deployCommand(clusterRepository: ClusterRepository, imageRepository: ImageRepository) {
    const question = new Question()
    try {
        console.log("Choose the target cluster")
        clusterRepository.clusters.forEach((cluster, index) => {
            console.log(chalk.bold.cyan(index.toString()) + ": " + pad(cluster, 5))
        })
        const targetCluster = await question.choose('> ', clusterRepository.clusters)

        const deployments = await load(clusterRepository.deployments(targetCluster))

        console.log("Choose the target deployment")
        deployments.forEach((deployment, index) => {
            console.log(chalk.bold.cyan(index.toString()) + ": " + pad(deployment.name, 5))
        })
        const targetDeployment = await question.choose('> ', deployments)

        const versions = await imageRepository.versions(targetDeployment.image!.name)
        versions.sort((a, b) => b.lastUpdated.getTime() - a.lastUpdated.getTime())
        console.log("Choose the target image")
        versions.forEach((image, index) => {
            if (image.tag === targetDeployment.image!.tag) {
                console.log(chalk.italic.magenta(index.toString()) + ": " + chalk.italic.magenta(pad(image.tag, 5)))
            } else {
                console.log(chalk.bold.cyan(index.toString()) + ": " + pad(image.tag, 5))
            }
        })
        const targetImage = await question.choose('> ', versions)

        console.log("deploy " + chalk.cyan(targetImage.url) + " into " + chalk.cyan(targetCluster + "::" + targetDeployment.name))
        console.log("Answer `yes` to apply.")
        const userAgrees = await question.expect('> ', "yes")
        if (userAgrees) {
            const updated = await clusterRepository.updateDeployment(targetCluster, targetDeployment, targetImage)
            logDeploymentHeader()
            logDeployment(updated)
        }
    } catch (error) {
        console.error(error)
    } finally {
        question.close()
    }
}

async function scalersCommand(clusterRepository: ClusterRepository, cluster: string) {
    const scalers = await load(clusterRepository.scalers(cluster))
    console.log(chalk.bold(pad("scaler name", 5) + pad("replication", 7)))
    console.log(chalk.bold(pad("", 5) + pad("current", 3) + pad("minimum", 2) + pad("maximum", 2)))
    scalers.forEach(scaler => {
        let line = pad(scaler.name, 5)
        let currentText = scaler.replicas.current.toString()
        if (scaler.replicas.current < scaler.replicas.minimum) {
            currentText = chalk.bold.cyan(pad(currentText + "!", 3))
        } else if (scaler.replicas.current >= scaler.replicas.maximum) {
            currentText = chalk.bold.magenta(pad(currentText, 3))
        } else {
            currentText = pad(currentText, 3)
        }
        line += currentText
        line += pad(scaler.replicas.minimum.toString(), 2) + pad(scaler.replicas.maximum.toString(), 2)
        console.log(line)
    })
}

export function printHelp(clusterRepository?: ClusterRepository, imageRepository?: ImageRepository, invalidConfig: boolean = false) {
    const b = chalk.blue;
    const m = chalk.magenta;
    const c = chalk.cyan;
    const CLI_TITLE = chalk.bold.underline('Spacegun-CLI')
    const CLI_DESCRIPTION = 'Space age deployment manager'
    const CLI_USAGE = 'Usage: \`spacegun <command> [options ...]\`'

    const HELP_HEADER = `
        ${b('/\\')} ${c('*')}    
       ${b('/__\\')}     ${CLI_TITLE}   ${b('version 0.0.8')}
      ${b('/\\  /')}
     ${b('/__\\/')}      ${CLI_DESCRIPTION}
    ${b('/\\')}  ${m('/\\')}     
   ${b('/__\\')}${m('/__\\')}     ${CLI_USAGE}
  ${b('/\\')}  ${m('/')}    ${m('\\')}
`
    console.log(HELP_HEADER)

    if (invalidConfig) {
        console.log(c('no configuration file found!'))
        console.log(c('A config.yml containing the following line might be sufficient'))
        console.log(b('docker: https://your.docker.registry/'))
        console.log("")
    }
    if (clusterRepository !== undefined) {
        console.log('configured clusters: ' + m(clusterRepository.clusters.join(", ")))
    } else {
        console.log(c('no clusters configured! (such as your kubernetes)'))
    }
    if (imageRepository !== undefined) {
        console.log('configured image endpoint: ' + m(imageRepository.endpoint))
    } else {
        console.log(c('no image registry configured! (such as your docker registry)'))
    }

    console.log('')
    console.log(chalk.bold.underline('Available Commands'))
    console.log(pad("pods", 2) + chalk.bold(pad("a summary of all pods of all known clusters", 10)))
    console.log(pad("images", 2) + chalk.bold(pad("a list of all images in the docker registry", 10)))
    console.log(pad("deployments", 2) + chalk.bold(pad("a summary of all deployements of all known clusters", 10)))
    console.log(pad("deploy", 2) + chalk.bold(pad("opens an interactive dialog to deploy an image", 10)))
    console.log(pad("scalers", 2) + chalk.bold(pad("a summary of all scalers of all known clusters", 10)))
    console.log(pad("help", 2) + chalk.bold(pad("renders this summary", 10)))
    console.log('')
    console.log(chalk.bold.underline('Available Options'))
    console.log(pad("config, c", 2) + chalk.bold(pad("path to the config.yml. Default: `config.yml`", 10)))
}

const commands: { [k in Command]: (clusterRepository: ClusterRepository, imageRepository: ImageRepository) => Promise<void> } = {
    "pods": async (clusterRepository: ClusterRepository, { }: ImageRepository) => {
        foreachCluster(clusterRepository, podsCommand)
    },
    "images": async ({ }: ClusterRepository, imageRepository: ImageRepository) => {
        imagesCommand(imageRepository)
    },
    "deployments": async (clusterRepository: ClusterRepository, { }: ImageRepository) => {
        foreachCluster(clusterRepository, deployementsCommand)
    },
    "scalers": async (clusterRepository: ClusterRepository, { }: ImageRepository) => {
        foreachCluster(clusterRepository, scalersCommand)
    },
    "deploy": async (clusterRepository: ClusterRepository, imageRepository: ImageRepository) => {
        deployCommand(clusterRepository, imageRepository)
    },
    "help": async (clusterRepository: ClusterRepository, imageRepository: ImageRepository) => {
        printHelp(clusterRepository, imageRepository)
    }
}

export { commands }
