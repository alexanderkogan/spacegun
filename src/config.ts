import { safeLoad } from "js-yaml"
import { readFileSync } from "fs"
import { homedir } from "os"

export interface Config {
    kube: string,
    docker: string
}

export function load(path: string = "./config.yml"): Config | Error {
    try {
        const doc = safeLoad(readFileSync(path, 'utf8')) as Partial<Config>
        if (doc.kube === undefined) {
            doc.kube = homedir() + "/.kube/config"
        }
        return doc as Config
    } catch (e) {
        return e
    }
}
