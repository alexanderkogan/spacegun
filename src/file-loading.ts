import { safeLoad, safeDump } from 'js-yaml'
import { readFileSync, writeFileSync, readdirSync } from 'fs'
import * as mkdirp from 'mkdirp'
import { parse } from 'path'

export function load(filePath: string): object {
    return safeLoad(readFileSync(filePath, 'utf8'))
}

export function save(filePath: string, data: object): Promise<void> {
    const path = parse(filePath)
    return new Promise((resolve, reject) => {
        mkdirp(path.dir, e => {
            if (e) {
                reject(e)
            } else {
                writeFileSync(filePath, safeDump(data, { skipInvalid: true }), 'utf8')
                resolve()
            }
        })
    })
}

export function list(filePath: string): string[] {
    return readdirSync(filePath)
}
