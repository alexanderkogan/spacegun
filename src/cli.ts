import { parse } from "./options"
import { KubernetesClusterProvider } from './cluster/kubernetes/KubernetesCluster'
import { commands } from "./commands"

(async () => {
    const options = parse()
    const cluster = new KubernetesClusterProvider(options.kube)
    try {
        await commands[options.command](cluster)
    } catch (error) {
        console.error(`Command ${options.command} failed: ${error}`)
    }
})()
