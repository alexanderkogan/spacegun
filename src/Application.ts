import { load as loadConfig, Config } from "@/config"
import { commands, printHelp } from "@/commands"
import { parse } from "@/options"

import { Layers } from "@/dispatcher/model/Layers"
import { run as runDispatcher } from "@/dispatcher"

import { IO } from "@/IO"
import { CronRegistry } from "@/crons/CronRegistry"

import { GitConfigRepository, fromConfig as gitRepoFromConfig } from "@/config/git/GitConfigRepository"
import { fromConfig as configRepoFromConfig } from "@/config/filesystem/FilesystemConfigRepository";

import { KubernetesClusterRepository } from "@/cluster/kubernetes/KubernetesClusterRepository"
import { DockerImageRepository } from "@/images/docker/DockerImageRepository"
import { JobsRepositoryImpl } from "@/jobs/JobsRepositoryImpl"
import { SlackEventRepository } from "@/events/slack/SlackEventRepository"

import { init as initCluster } from "@/cluster/ClusterModule"
import { init as initEvents } from "@/events/EventModule"
import { init as initImages } from "@/images/ImageModule"
import { init as initConfig } from "@/config/ConfigModule"
import { init as initJobs } from "@/jobs/JobsModule"
import { init as initViews } from "@/views"

import { Options } from "@/options"

export class Application {

    public static create(): Application {
        const io = new IO()
        const crons = new CronRegistry()
        const options = parse()
        return new Application(io, crons, options)
    }

    public constructor(
        public readonly io: IO,
        public readonly crons: CronRegistry,
        public readonly options: Options
    ) { }

    public async run() {
        try {
            const config = loadConfig(this.options.config)
            this.initialize(config)

            runDispatcher(config.server.host, config.server.port)
            if (process.env.LAYER === Layers.Standalone || process.env.LAYER === Layers.Client) {
                await commands[this.options.command](this.io)
            }
        } catch (e) {
            printHelp(this.io, e)
        }
    }

    public async checkForConfigChange(git: GitConfigRepository) {
        const hasNewConfig = await git.hasNewConfig()
        if (hasNewConfig) {
            await git.fetchNewConfig()
            this.crons.removeAllCrons()
            this.reload()
        }
    }

    private reload(): void {
        try {
            const config = loadConfig(this.options.config)
            this.initialize(config)
        } catch (e) {
            this.io.out(`could not reload config ${e.message}`)
        }
    }

    private initialize(config: Config) {
        const gitRepo = gitRepoFromConfig(config)
        if (gitRepo !== undefined) {
            this.crons.register(
                "config-reload",
                config.git!.cron,
                () => this.checkForConfigChange(gitRepo)
            )
            initConfig(gitRepo)
        } else {
            const fileSystemRepo = configRepoFromConfig(config)
            initConfig(fileSystemRepo)
        }

        initViews(config)
        initEvents([
            SlackEventRepository.fromConfig(config.slack)
        ])
        initCluster(KubernetesClusterRepository.fromConfig(config.kube, config.namespaces))
        initImages(DockerImageRepository.fromConfig(config.docker))
        initJobs(JobsRepositoryImpl.fromConfig(config.jobs, this.crons))
    }
}
