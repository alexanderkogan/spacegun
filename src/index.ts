import { Layers } from "@/dispatcher/model/Layers"

import { load as loadConfig } from "@/config"
import { parse } from "@/options"
import { run as runDispatcher } from "@/dispatcher"
import { commands, printHelp } from "@/commands"
import { IO } from "@/IO"

import { init as initJobs } from "@/jobs/JobsModule"
import { init as initCluster } from "@/cluster/ClusterModule"
import { init as initImages } from "@/images/ImageModule"

import { KubernetesClusterRepository } from "@/cluster/kubernetes/KubernetesClusterRepository"
import { DockerImageRepository } from "@/images/docker/DockerImageRepository"
import { JobsRepositoryImpl } from "@/jobs/JobsRepositoryImpl"

const options = parse()
const config = loadConfig(options.config)
const io = new IO()
try {
    if (config instanceof Error) {
        printHelp(io, true)
    } else {
        initJobs(JobsRepositoryImpl.fromConfig(config.jobs))
        initCluster(KubernetesClusterRepository.fromConfig(config.kube))
        initImages(DockerImageRepository.fromConfig(config.docker))

        runDispatcher(config.server.host, config.server.port)
        if (process.env.LAYER === Layers.Standalone || process.env.LAYER === Layers.Client) {
            commands[options.command](io)
        }
    }
} catch (e) {
    console.error(e)
}
