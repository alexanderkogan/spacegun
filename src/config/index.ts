import { homedir } from 'os'
import { parse } from 'path'

import { load } from '../file-loading'

export interface ServerConfig {
    host: string
    port: number
}

export interface GitConfig {
    remote: string
    cron: string
}

export interface Config {
    kube: string
    docker: string
    pipelines: string
    artifacts: string
    configBasePath: string
    slack?: string
    git?: GitConfig
    server: ServerConfig
    namespaces?: string[]
}

export function loadConfig(filePath: string = './config.yml'): Config {
    const path = parse(filePath)
    const doc = load(filePath) as Partial<Config>
    return validateConfig(path.dir, doc)
}

export function validateConfig(configBasePath: string, partial: Partial<Config>): Config {
    let { kube, namespaces, pipelines, artifacts, slack, server = {}, docker, git } = partial

    if (docker === undefined) {
        throw new Error(`a docker endpoint is needed`)
    }

    if (kube === undefined) {
        kube = homedir() + '/.kube/config'
    } else {
        kube = `${configBasePath}/${kube}`
    }

    if (pipelines === undefined) {
        pipelines = `${configBasePath}/pipelines`
    } else {
        pipelines = `${configBasePath}/${pipelines}`
    }

    if (artifacts === undefined) {
        artifacts = `${configBasePath}/artifacts`
    } else {
        artifacts = `${configBasePath}/${artifacts}`
    }

    return { configBasePath, kube, pipelines, artifacts, slack, namespaces, git, server: validateServerConfig(server), docker }
}

export function validateServerConfig(partial: Partial<ServerConfig>): ServerConfig {
    const { port = Number.parseInt(process.env.SERVER_PORT || '3000'), host = process.env.SERVER_HOST || 'http://localhost' } = partial

    return { port, host }
}
