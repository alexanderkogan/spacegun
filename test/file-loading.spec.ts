import { load, save } from "../src/file-loading"

const mockReadFile = jest.fn()
const mockWriteFile = jest.fn()
const mockMkdir = jest.fn()
let mockMkdirError: Error | undefined = undefined

jest.mock("fs", () => ({
    readFileSync: (path: string) => mockReadFile(path),
    writeFileSync: (path: string, data: string) => mockWriteFile(path, data)
}))

jest.mock("mkdirp", () => ((path: string, callback: (e?: Error) => void) => {
    mockMkdir(path)
    callback(mockMkdirError)
}))

const ymlSerialObject = "some: param\nsomeOther: 2\n"
const testObject = { some: "param", someOther: 2 }

describe(load.name, () => {

    it("loads yaml files", () => {
        mockReadFile.mockReturnValue(ymlSerialObject)

        const file = load("test/test-config/config.yml")

        expect(file).toEqual(testObject)
        expect(mockReadFile).toHaveBeenCalledWith("test/test-config/config.yml")
    })
})

describe(save.name, () => {

    it("saves yaml files", async () => {
        await save("some/path/file.yml", testObject)

        expect(mockMkdir).toHaveBeenCalledWith("some/path")
        expect(mockWriteFile).toHaveBeenCalledWith("some/path/file.yml", ymlSerialObject)
    })

    it("propagates file io errors", () => {
        mockMkdirError = new Error("some issue happened")

        return expect(save("some/path/file.yml", testObject)).rejects.toMatchSnapshot()
    })
})
