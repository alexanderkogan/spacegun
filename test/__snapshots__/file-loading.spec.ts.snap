// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`save propagates file io errors 1`] = `[Error: some issue happened]`;
